### PR TITLE
Layout animations for container array'd objects

### PR DESCRIPTION
**Pull request by rreusser**
This PR does two things:

1. a couple one-line tweaks were needed to get layouts to use the new `extendLayout` function from #1041 
2. transitions with `redraw: true` weren't getting redrawn between frames due to a race condition between frame duration and transition duration. If you specify `redraw: true`, the whole plot now *definitely* gets redrawn between frames.

Here are animated annotations. Note that they take effect at the *end* of the transition since smooth annotation transitions are not yet implemented. Also, two-way component binding is *not* merged into this branch but will work once #1016 is merged.

annotation-animation

### CODE REVIEW

**Comment by rreusser** [comment]
Adding a test for annotation animation example.

**Comment by rreusser** [comment]
Added test. Sorry for the messy git history. The linter strikes again.

File: src/plots/plots.js [hunk 1]
```diff
@@ -1558,10 +1558,25 @@ plots.transition = function(gd, data, layout, traces, frameOpts, transitionOpts)
...
+            // If another transition is triggered, this callback will be executed simply because it's
+            // in the interruptCallbacks queue. If this transition completes, it will instead flush
+            // that queue and forget about this callback.
             gd._transitionData._interruptCallbacks.push(function() {
                 aborted = true;
             });
 
+            if(frameOpts.redraw) {
+                gd._transitionData._interruptCallbacks.push(function () {
+                    return Plotly.redraw(gd);
```
Comment (etpinard) [inline src/plots/plots.js]: nice use of `Plotly.redraw`

File: test/jasmine/tests/transition_test.js [hunk 2]
```diff
@@ -65,6 +65,34 @@ function runTests(transitionDuration) {
...
+            }).then(function () {
+                p2 = annotationPosition();
+
+                // Ensure both coordinates have moved, i.e. that the annotation has transitioned:
+                expect(p1[0]).not.toEqual(p2[0]);
+                expect(p1[1]).not.toEqual(p2[1]);
+
+            }).catch(fail).then(done);
+
+        });
```
Comment (etpinard) [inline test/jasmine/tests/transition_test.js]: Thanks!

File: src/plots/plots.js [hunk 3]
```diff
@@ -1510,7 +1512,7 @@ plots.transition = function(gd, data, layout, traces, frameOpts, transitionOpts)
             delete layoutUpdate[attr].range;
         }
 
-        Lib.extendDeepNoArrays(gd.layout, layoutUpdate);
+        plots.extendLayout(gd.layout, layoutUpdate);
```
Comment (etpinard) [inline src/plots/plots.js]: so, this now guarantees that `shapes`, `images` and all the other layout array container can also be animated?

Comment (rreusser) [inline src/plots/plots.js]: Unless there are bugs, yes. It iterates through each of those container arrays as a way of bypassing the `No` in `extendDeepNoArrays`. The option to force redraw makes non-animatable things Just Work.

Comment (etpinard) [inline src/plots/plots.js]: Great. Let's :lock: this down by expanding test case 

https://github.com/plotly/plotly.js/pull/1081/files#diff-2032f9a39fb6bbc9611ed5ec74c10c19R68

to transition all layout array containers.

Comment (rreusser) [inline src/plots/plots.js]: Lunch, then adding it.

**Comment by rreusser** [comment]
@etpinard Images took just a bit of tweaking to preserve the image element in the join. Flickered excessively otherwise...

cat

**Comment by rreusser** [comment]
@etpinard Added shape and image tests. Ensured that the correct properties are modified or remain the same. Nothing too fancy or profound. Just some sanity checks. Both still need *actual* animation (i.e. smooth transition) support, but this is a good start.

**Comment by etpinard** [comment]
amazing :dancer:

**Comment by etpinard** [comment]
@rreusser follow up question on this PR: can frames clear array containers (e.g. annotations) on `.animate`? See example: http://codepen.io/etpinard/pen/WoNryW

**Pull request merged**